app/domain/authentication/authn_k8s/execute_command_in_container.rb

```diff
@@ -119,11 +109,41 @@ def on_error(err)
 
       private
 
-      def add_websocket_event_handlers(ws_client, body, stdin)
+      def init_ws_client
+        @message_log    = @message_log_class.new
+        @channel_closed = false
+        ws_client
+      end
+
+      def ws_client
+        @ws_client ||= @websocket_client.connect(server_url, headers: headers)
+      end
+
+      def server_url
+        api_uri  = kube_client.api_endpoint
+        base_url = "wss://#{api_uri.host}:#{api_uri.port}"
+        path     = "/api/v1/namespaces/#{@pod_namespace}/pods/#{@pod_name}/exec"
+
+        base_query_string_parts = %W(container=#{CGI.escape(@container)} stderr=true stdout=true)
+        stdin_part              = @stdin ? ['stdin=true'] : []
+        cmds_part               = @cmds.map { |cmd| "command=#{CGI.escape(cmd)}" }
+        query_string            = (base_query_string_parts + stdin_part + cmds_part).join("&")
+
+        "#{base_url}#{path}?#{query_string}"
+      end
+
+      def headers
+        @headers ||= kube_client.headers.clone
+      end
+
+      def add_websocket_event_handlers(ws_client)
         # We need to set this so the handlers will call this class's methods.
         # If we use 'self' inside the curly brackets it will be try to use methods
-        # of the class WebSocket::Client::Simple::Client
-        kube = self
+        # of the class WebSocket::Client::Simple::Client.
+        # We need to have method params for the same reason.
+        kube  = self
+        body  = @body
+        stdin = @stdin
```

Comment:
It's not necessary to make do `kube = self` and make the `on_XXX` methods of this CommandClass public.  It can and should have the single public `call` method like all CommandClass's.  Ruby can create a closure over a private method with no issues:

[Demo of closing over a private method](https://tio.run/##VZBBCoMwEEX3nmJIoejGAxQKQi/QRfclJmMbGqMk0dqKZ09johY3YWb@48@f6K78OMckNQYuUqCyCQDHChp1x9636bGUr8wPAQpfwRn861tUPFlQRqXcgHzpZj0w0fuGZnUWSlhBpfhiysLGxX3w3sQ0NUKrRU8tAqeWkiBGMF9Dwbgy9xrts@Ew/SMtyppuDwa3trMGyFULZYV6QKWbetsZsRMcxmKYyO4Qny9@Ua7wncwXzUXKsoSFq537AQ "Ruby – Try It Online")

Reply:
I have added this fix to my branch.

Reply:
after this change we get: `undefined method `on_error' for #<WebSocket::Client::Simple::Client:0x00005589935f0060> (NoMethodError)`